### PR TITLE
Fix ArrayAgg schema mismatch issue

### LINKED ISSUE #8032: array_agg with pyarrow errors with ArrowInvalid: Schema at index 0 was different
### Describe the bug

When using the `array_agg` function in a query on a DataFusion table generated by a pyarrow Dataset on parquet files, the .to_pandas() and .to_polars() functions error with the below stacktrace:

```
---------------------------------------------------------------------------
ArrowInvalid                              Traceback (most recent call last)
Cell In[29], line 1
----> 1 results.to_pandas()

File /opt/conda/envs/opt-strategies-amer/lib/python3.10/site-packages/pyarrow/table.pxi:4057, in pyarrow.lib.Table.from_batches()

File /opt/conda/envs/opt-strategies-amer/lib/python3.10/site-packages/pyarrow/error.pxi:154, in pyarrow.lib.pyarrow_internal_check_status()

File /opt/conda/envs/opt-strategies-amer/lib/python3.10/site-packages/pyarrow/error.pxi:91, in pyarrow.lib.check_status()

ArrowInvalid: Schema at index 0 was different: 
double_field: list<item: double>
string_field: list<item: string>
vs
double_field: list<item: double> not null
string_field: list<item: string> not null
```

### To Reproduce

Using latest pyarrow and datafusion (other versions not tested):
pyarrow==14.0.0
datafusion==32.0.0

```python
import pandas as pd
df = pd.DataFrame({'double_field': [1.5, 2.5, 3.5], 'string_field': ['a', 'b', 'c']})
df.to_parquet("test_array_agg.parquet")

from pyarrow import dataset
test_dataset = dataset.dataset("test_array_agg.parquet")

from datafusion import SessionContext
ctx = SessionContext()
ctx.register_dataset("test_table", test_dataset)
tbl = ctx.table("test_table")

query = """
SELECT
    array_agg("double_field" ORDER BY "string_field") as "double_field",
    array_agg("string_field" ORDER BY "string_field") as "string_field"
FROM test_table
"""
results = ctx.sql(query)

# Result schema is list<item: double>
print(results.schema())

# First pyarrow.RecordBatch schema is list<item: double> not null
print(results.collect()[0].schema)

# ERRORS with ArrowInvalid
results.to_pandas()
```

### Expected behavior

The `.to_pandas()` method should return a DataFrame with arrays in each cell.

### Additional context

The query works and returns the expected result if you wrap array_agg in additional calls to make_array and array_extract:

```python
SELECT
    array_extract(make_array(array_agg("double_field") ORDER BY "string_field"), 1) as "double_field",
    array_extract(make_array(array_agg("string_field") ORDER BY "string_field"), 1) as "string_field"
FROM test_table
```

Returns from .to_pandas() with:

image

In this case both `results.schema()` and `results.collect()[0].schema` are `list<item: double>`/`list<item: string>`

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #8032.

## Rationale for this change

The main issue for #8032 is as what the errors said we have different schema for result.schema which is `logical-expr schema` and results.collect()[0].schema which is `physical-expr schema`.

We should have the same `is_nullable` for both logical-expr and physical-expr

We can see that the logical-expr of List is set with expr.nullable(schema)
https://github.com/apache/arrow-datafusion/blob/3469c4e09a3d32381949dd0c0f626f406c00c6ad/datafusion/expr/src/expr_schema.rs#L289-L305
and logical-expr of List element is set with the return type of Aggregatefunciton::ArrayAgg
https://github.com/apache/arrow-datafusion/blob/3469c4e09a3d32381949dd0c0f626f406c00c6ad/datafusion/expr/src/aggregate_function.rs#L286-L290

## What changes are included in this PR?

Fix schema `is_nullable`

## Are these changes tested?

I did not test end to end like #8032 , just make sure the schema is the same.

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/array_agg.rs [hunk 1]
```diff
@@ -195,6 +198,30 @@ mod tests {
...
+        ($ARRAY:expr, $DATATYPE:expr, $OP:ident, $EXPECTED:expr, $EXPECTED_DATATYPE:expr) => {{
+            let schema = Schema::new(vec![Field::new("a", $DATATYPE, true)]);
+
+            let batch = RecordBatch::try_new(Arc::new(schema.clone()), vec![$ARRAY])?;
+
+            let agg = Arc::new(<$OP>::new(
+                col("a", &schema)?,
+                "bla".to_string(),
+                $EXPECTED_DATATYPE,
+                true,
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/aggregate/array_agg.rs:214]: The only different from generic_test_op!()

File: datafusion/physical-expr/src/aggregate/array_agg_distinct.rs [hunk 2]
```diff
@@ -48,12 +50,14 @@ impl DistinctArrayAgg {
         expr: Arc<dyn PhysicalExpr>,
         name: impl Into<String>,
         input_data_type: DataType,
+        is_expr_nullable: bool,
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/array_agg_distinct.rs:53]: should we perhaps update the docstrings here to explain `is_expr_nullable` means if the input can have nulls?

**Review by alamb — APPROVED** [review]
Thank you @jayzhan211  --  it would be nice if there was some way to avoid having to plumb the `is_expr_nullable` flag through, but I don't think there is

File: datafusion/physical-expr/src/aggregate/array_agg.rs [hunk 3]
```diff
@@ -37,6 +37,7 @@ pub struct ArrayAgg {
     name: String,
     input_data_type: DataType,
     expr: Arc<dyn PhysicalExpr>,
+    is_expr_nullable: bool,
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/array_agg.rs:40]: What do you think about calling this just `nullable` ? it is a bit more concise and aligns with the name used in `Field`

**Review by alamb — COMMENTED** [review]
Thank you @jayzhan211  for the contribution and working on this area of the code

**Review by alamb — APPROVED** [review]
Thanks again @jayzhan211

**Pull request merged**